plugins/SpectrumAnalyzer/SaProcessor.cpp

```diff
@@ -0,0 +1,556 @@
+	m_bufferL.resize(m_fftBlockSize, 0);
+	m_bufferR.resize(m_fftBlockSize, 0);
+	m_spectrumL = (fftwf_complex *) fftwf_malloc(binCount() * sizeof (fftwf_complex));
+	m_spectrumR = (fftwf_complex *) fftwf_malloc(binCount() * sizeof (fftwf_complex));
+	m_fftPlanL = fftwf_plan_dft_r2c_1d(m_fftBlockSize, m_bufferL.data(), m_spectrumL, FFTW_MEASURE);
+	m_fftPlanR = fftwf_plan_dft_r2c_1d(m_fftBlockSize, m_bufferR.data(), m_spectrumR, FFTW_MEASURE);
+
+	m_absSpectrumL.resize(binCount(), 0);
+	m_absSpectrumR.resize(binCount(), 0);
+	m_normSpectrumL.resize(binCount(), 0);
+	m_normSpectrumR.resize(binCount(), 0);
+
+	m_history.resize(binCount() * m_waterfallHeight * sizeof qRgb(0,0,0), 0);
+
+	clear();
+}
+
+
+SaProcessor::~SaProcessor()
+{
+	// take a note that destructor was called; see reallocateBuffers()
+	m_destroyed = true;
+
+	if (m_fftPlanL != NULL) {fftwf_destroy_plan(m_fftPlanL);}
+	if (m_fftPlanR != NULL) {fftwf_destroy_plan(m_fftPlanR);}
+	if (m_spectrumL != NULL) {fftwf_free(m_spectrumL);}
+	if (m_spectrumR != NULL) {fftwf_free(m_spectrumR);}
+
+	m_fftPlanL = NULL;
+	m_fftPlanR = NULL;
+	m_spectrumL = NULL;
+	m_spectrumR = NULL;
+}
+
+
+// Load a batch of data from LMMS; run FFT analysis if buffer is full enough.
+void SaProcessor::analyse(sampleFrame *in_buffer, const fpp_t frame_count)
+{
+	#ifdef SA_DEBUG
+		int start_time = std::chrono::high_resolution_clock::now().time_since_epoch().count();
+	#endif
+	// only take in data if any view is visible and not paused
+	if ((m_spectrumActive || m_waterfallActive) && !m_controls->m_pauseModel.value())
+	{
+		const bool stereo = m_controls->m_stereoModel.value();
+		fpp_t in_frame = 0;
+		while (in_frame < frame_count)
+		{
+			// fill sample buffers and check for zero input
+			bool block_empty = true;
+			for (; in_frame < frame_count && m_framesFilledUp < m_inBlockSize; in_frame++, m_framesFilledUp++)
+			{
+				if (stereo)
+				{
+					m_bufferL[m_framesFilledUp] = in_buffer[in_frame][0];
+					m_bufferR[m_framesFilledUp] = in_buffer[in_frame][1];
+				}
+				else
+				{
+					m_bufferL[m_framesFilledUp] = (in_buffer[in_frame][0] + in_buffer[in_frame][1]) * 0.5;
+					m_bufferR[m_framesFilledUp] = (in_buffer[in_frame][0] + in_buffer[in_frame][1]) * 0.5;
+				}
+				if (in_buffer[in_frame][0] != 0 || in_buffer[in_frame][1] != 0)
+				{
+					block_empty = false;
+				}
+			}
+	
+			// Run analysis only if buffers contain enough data.
+			// Also, to prevent audio interruption and a momentary GUI freeze,
+			// skip analysis if buffers are being reallocated.
+			if (m_framesFilledUp < m_inBlockSize || m_reallocating) {return;}
+
+			// update sample rate
+			m_sampleRate = Engine::mixer()->processingSampleRate();
+	
+			// apply FFT window
+			for (int i = 0; i < m_inBlockSize; i++)
+			{
+				m_bufferL[i] = m_bufferL[i] * m_fftWindow[i];
+				m_bufferR[i] = m_bufferR[i] * m_fftWindow[i];
+			}
+	
+			// lock data shared with SaSpectrumView and SaWaterfallView
+			QMutexLocker lock(&m_dataAccess);
```

Comment:
I think I understand only about 50 % of what you say. :D
I don't have much experience with multi-threaded applications, so I have no clue what it means that LMMS is not realtime safe. Mutex lock seems like a straight-forward and appropriate tool for the job, so I just used that (at that point I did not even know something like a mutex locker even existed).
It seems to me that using atomic variables is just a way how to implement locking yourself, in a more complex way? I mean, what do I even do? Wait in a loop until the atomic variable changes state? That does not seem quite right..

Reply:
So, today I finished reading one of the articles you suggested earlier. And while I'm now convinced that having locks in audio threads is probably not a good idea, I'm now even more confused about possible solutions -- it seems there is very little consensus about "the right way to do it", or rather, there is no standard and widely supported way to do so and most people simply do not care, because they do not work with audio or other latency sensitive workloads.

There are some random third party libraries and implementations, and also articles warning you that even a simple and obvious implementation of a FIFO or ring buffer is not easy to get right, so I'm pretty sure I have no business designing my own, I don't have nearly enough experience for that.

So it seems to me the best way would be to use one of the published implementations -- but that's not something that should happen in this PR, it should be discussed and decided in the wider context of "making LMMS real-time safe", because it will make sense to reuse the implementation in other parts of LMMS. There seem to be even issues like software patents around lock-less algorithms, so it probably warrants some deeper thought.

To finish this PR, I thought of just using std::queue instead of the "safe FIFO" to design the lock-less interface between SaProcessor and Sa###View classes. It could work as a place-holder and be later replaced with the proper implementation, but in the mean time it could potentially cause problems, so I'm not sure it would be a good thing, compared to the current safe version that uses locks.

Reply:
> I don't have nearly enough experience for that

Don't say that, that stuff is actually suited well for teaching parallel computing. Btw, there may be a "Google Summer of Code" project in 2020 where we make LMMS realtime safe, if you are a student, you might even participate.

It's a good point that ringbuffers should be a global decision and put into the core, rather than us doing our own thing for one plugin. I once wrote my own [ringbuffer library](https://github.com/JohannesLorenz/ringbuffer) for that (it's partially a copy of JACK's ringbuffer). It has tests, and I use it in daily audio production, so it's safe. I might start an issue putting this into the LMMS core, but that stuff can take a bit time to review. Thanks for reminding me of that.

Per an `std::queue` ringbuffer... not such a bad idea. You could write a mini-class and start with that. If you want that, I agree that this should go into another branch, to not damage the current safe version. But there's no need to hurry now, only try it if you want.

Reply:
> that stuff is actually suited well for teaching parallel computing.

Is it? To me, it seems it goes a bit beyond the basic concepts taught at school. I'm not a student anymore, but almost all my practical experience is with embedded systems and FPGA development. That is, in C I only did low-level single-core stuff and the first time I even used a mutex in practice was in this plugin. :)

After reading some articles, I'm now aware of another layer of concurrency problems that I did not even consider before -- there is the CPU instructions reordering (which I knew about, but did not consider in this context), there is the problem of each core having its own cache that takes time to be synchronized, then there is the problem with non-x86 architectures that will again behave differently and will possibly break a solution you thought was OK, and then there are other possible problems I'm not even aware of.

So, yes, I think I'm pretty inexperienced in this regard. I'm at the stage where I know many of these problems exist, but I have no experience solving them or even finding them. And if I try to do so, the result will be probably quite poor and full of hidden bugs or problems -- i.e. not suited for use in a program used by [many] people.

But that's all just to say I prefer using already existing and well tested library or implementation instead (and we seem to be in agreement on that).